src/coreclr/src/jit/indirectcalltransformer.cpp

```diff
@@ -760,6 +789,179 @@ class IndirectCallTransformer
+            GenTree*                 nullCheck = argsIter.GetUse()->GetNode();
+            ++argsIter;
+            GenTree* sizeCheck = argsIter.GetUse()->GetNode();
+            ++argsIter;
+            origCall->gtCallArgs = argsIter.GetUse();
+            // The first argument is the handle now.
+            checkBlock = CreateAndInsertBasicBlock(BBJ_COND, currBlock);
+
+            assert(sizeCheck->OperIs(GT_LE));
+            GenTree*   sizeJmpTree = compiler->gtNewOperNode(GT_JTRUE, TYP_VOID, sizeCheck);
+            Statement* sizeJmpStmt = compiler->fgNewStmtFromTree(sizeJmpTree, stmt->GetILOffsetX());
+            compiler->fgInsertStmtAtEnd(checkBlock, sizeJmpStmt);
+
+            checkBlock2 = CreateAndInsertBasicBlock(BBJ_COND, checkBlock);
+            assert(nullCheck->OperIs(GT_EQ));
+            GenTree*   nullJmpTree = compiler->gtNewOperNode(GT_JTRUE, TYP_VOID, nullCheck);
+            Statement* nullJmpStmt = compiler->fgNewStmtFromTree(nullJmpTree, stmt->GetILOffsetX());
+            compiler->fgInsertStmtAtEnd(checkBlock2, nullJmpStmt);
+        }
+
+        //------------------------------------------------------------------------
+        // CreateThen: create then block, that is executed if the checks succeed.
+        // This simply returns the handle.
+        //
+        virtual void CreateThen() override
+        {
+            thenBlock = CreateAndInsertBasicBlock(BBJ_ALWAYS, checkBlock2);
+
+            GenTreeCall::UseIterator argsIter     = origCall->Args().begin();
+            GenTree*                 resultHandle = argsIter.GetUse()->GetNode();
+            ++argsIter;
+            // The first argument is the real first argument for the call now.
+            origCall->gtCallArgs = argsIter.GetUse();
+
+            GenTree*   asg     = compiler->gtNewTempAssign(resultLclNum, resultHandle);
+            Statement* asgStmt = compiler->gtNewStmt(asg, stmt->GetILOffsetX());
+            compiler->fgInsertStmtAtEnd(thenBlock, asgStmt);
+        }
+
+        //------------------------------------------------------------------------
+        // CreateElse: create else block, that is executed if the checks fail.
+        //
+        virtual void CreateElse() override
+        {
+            elseBlock          = CreateAndInsertBasicBlock(BBJ_NONE, thenBlock);
+            GenTree*   asg     = compiler->gtNewTempAssign(resultLclNum, origCall);
+            Statement* asgStmt = compiler->gtNewStmt(asg, stmt->GetILOffsetX());
+            compiler->fgInsertStmtAtEnd(elseBlock, asgStmt);
+        }
+
+        //------------------------------------------------------------------------
+        // SetWeights: set weights for new blocks.
+        //
+        virtual void SetWeights() override
+        {
+            remainderBlock->inheritWeight(currBlock);
+            checkBlock->inheritWeight(currBlock);
+            // TODO: fix weights here based on real values, ask Fadi.
```

Comment:
What are the weights you are looking for here?

Reply:
How often (in percentages) will size check fail? How often will null check fail?

We use these weights for block layout (maybe for cold/hot code splitting in the future) when we don't have real dynamic counters.

Reply:
These checks are one-time initialization. The elseBlock will be typically executed once (or very few times), and both the checks will be succeeding after that.